QEP 343: Customized Toolbars and Menus (#343)
qgis/QGIS-Enhancement-Proposals: qep-343-customized-toolbars-and-menus.md

Additions to the Interface Customization dialog would be:

  * similarly to the attribute form settings, a drop-down that allows to use a "drag and drop designer" for the configuration of the different components of the GUI
  * if the "drag and drop designer" is chosen, the possibility (green plus icon) to add new "containers" (similarly to adding new containers in the attribute form settings), with the choice between "menu" and "toolbar"
  * a "provide .ini file" option to load an existing configuration
  * an "export configuration" button opening the profile folder at QGIS > QGISCUSTOMIZATION.ini

![](./images/qep343/drop-down-menu.png)

Possible integration of a drop-down menu similar to the layer attribute form dialog

![](./images/qep343/customize-toolbars-plugin.png)

Current interface of the Customize ToolBars plugin

![](./images/qep343/drag-drop-designer-container.png)

Layer attribute form dialog (drag and drop designer; add container)

### Delimitation

>   * A restart of QGIS is necessary to take into account the changes (as is already the case for modifications in the Interface Customization dialog).

Author: QEP updated accordingly.

>   * Processing algorithm features are only available if they have been added to the toolbars beforehand through the Settings > Options > Processing > Menus dialog → activate "Add button in toolbar"

Reviewer: Could we drop that whole processing feature and just move it to the new toolbar customisation function? I don't really like how it's implemented at the moment and would love to see processing settings de-cluttered.

>   * Processing algorithm features are only available if they have been added to the toolbars beforehand through the Settings > Options > Processing > Menus dialog → activate "Add button in toolbar"

Author: Yes, I agree, the current implementation is not satisfying, and we can take the opportunity to move this feature in the new customization logic.

>   * Processing algorithm features are only available if they have been added to the toolbars beforehand through the Settings > Options > Processing > Menus dialog → activate "Add button in toolbar"

Author: I updated the QEP accordingly

  * If plugin actions are added to custom toolbars/menus and the configuration is imported within a profile where those plugins are missing, the missing items will be ignored. There will be no tracking of dependencies from plugins and therefore no warnings if a certain action is missing.
  * Personal Python scripts will not be available.
  * Customization is limited to toolbars and menus, as we have not identified any need for further customization of the browser, docks, status bar or widgets so far.

## Risks

None known.

## Performance Implications

No change in performance to be expected.

## Further Considerations/Improvements

> In a further step, the integration of all processing algorithms to choose from for custom menus or toolbars could be considered. Any additional need for customization for browser, dock, status bar or widget elements could also be discussed.

Reviewer 2: If I understand well the `Personal Python scripts will not be available.` part, perhaps @Djedouas is also referring to Processing algorithms written as Python scripts, which wouldn't be included in the initial proposal.

> In a further step, the integration of all processing algorithms to choose from for custom menus or toolbars could be considered. Any additional need for customization for browser, dock, status bar or widget elements could also be discussed.

Author: QEP adapted accordingly

  * the storing of the settings in a binary file outside the QGIS folder structure
  * possible maintenance issues
  * loading of saved configurations is not possible through the plugin GUI (less intuitive for users)
  * it's a plugin

## Proposed Solution

The integration of customized toolbars within the Interface Customization dialog would be a logical follow-up to the already available GUI customization possibilities. The configuration would also be saved within the user profile QGIS > QGISCUSTOMIZATION3.ini file.

Additions to the Interface Customization dialog would be:

  * similarly to the attribute form settings, a drop-down that allows to use a "drag and drop designer" for the configuration of the different components of the GUI
  * if the "drag and drop designer" is chosen, the possibility (green plus icon) to add new "containers" (similarly to adding new containers in the attribute form settings), with the choice between "menu" and "toolbar"
  * a "provide .ini file" option to load an existing configuration
  * an "export configuration" button opening the profile folder at QGIS > QGISCUSTOMIZATION.ini

> 

Reviewer 3: Do you think it will be possible to import an existing toolbar[s] (clone) to adjust it[them] to user needs?

Author: Good point, I think that in the UI where the user selects the tool-buttons he wants to add to his custom toolbar, he will browse existing tool-buttons and existing toolbars.

If he selects an existing toolbar, it will add all the buttons at once in his customized toolbar.

Author: QEP updated accordingly

> ![](./images/qep343/drop-down-menu.png)
> 
> Possible integration of a drop-down menu similar to the layer attribute form dialog
> 
> ![](./images/qep343/customize-toolbars-plugin.png)
> 
> Current interface of the Customize ToolBars plugin
> 
> ![](./images/qep343/drag-drop-designer-container.png)
> 
> Layer attribute form dialog (drag and drop designer; add container)

Reviewer 3: Suggestion to improve the readability: 

```suggestion
Possible integration of a drop-down menu similar to the layer attribute form dialog:
![](./images/qep343/drop-down-menu.png)

Current interface of the Customize ToolBars plugin:
![](./images/qep343/customize-toolbars-plugin.png)

Layer attribute form dialog (drag and drop designer; add container):
![](./images/qep343/drag-drop-designer-container.png)
```

### Processing algorithms

Currently, processing algorithms can be added to the "Processing Algorithms Toolbar", through a non-intuitive and cumbersome user experience.

> We propose to drop this logic, to integrate all processing algorithms in the new toolbars and menus customization.

Reviewer 3: At first, I believed you want to add all processing to any new customisation! Here is my suggestion:
```suggestion
To replace this logic, we propose to have the possibility of integrating any processing algorithm into the new customization of toolbars and menus.
```

### Affected Files

- src/app/qgscustomization.[cpp|h]

The new interface to customize toolbars and menus will be new sections in the Interface Customization dialog:

> ![](./images/qep343/proposal-settings.png)
> 
> Example of new sections in the Interface Customization dialog.

Reviewer 3: ```suggestion
![Example of new sections in the Interface Customization dialog](./images/qep343/proposal-settings.png)
```

### Delimitation

>   * A restart of QGIS could be necessary to take into account the changes (as it is already the case for modifications in the Interface Customization dialog).

Reviewer 4: >  A restart of QGIS could be necessary to take into account the changes (as it is already the case for modifications in the Interface Customization dialog).

This is a very big limitation. Unlike the current interface customization feature which is mainly targeted to admins creating qgis deployments, this one should be targeting average end users. Creating custom toolbars should be as simple as moving them around and toggling them on and off

Author: I agree that when we decide if this feature should require a restart or not, we have to include this decision in this initial QEP.
